### DISCUSSION TITLE
How to figure out M2M column name from meta data

### DISCUSSION THREAD

**Discussion by sclu1034** [nocodb/nocodb#7971] [Q&A]
I have two tables `Songs` & `Playlists`. Each have a column that points to the other, with a M2M relationship.

When querying `Playlists` via the API, I get a field `Songs` (which matches the name in the UI), but its content is just a number. The field holding the actual data has some generated name `nc_o9qt___nc_m2m_odp0l09uz_s`.

Since I haven't found any documentation, I've already been digging through the source code and found a few "meta" API routes, but especially the table meta contains quite a lot of stuff.

What series of API calls and their combination do I need to figure out the JSON field name for a M2M column, assuming that the only info I have are the human-readable strings in the UI (i.e. table names, column names)?

**Comment by mertmit** [comment]
Hi @sclu1034,
The number you are seeing is reflecting how many records are linked within that column.

If you would like to see some specific information related to linked records you can use a Lookup column (https://docs.nocodb.com/fields/field-types/links-based/lookup) or Rollup column for aggregated information (https://docs.nocodb.com/fields/field-types/links-based/rollup/).

If you would like to get linked records via API you can use `/api/v2/tables/{tableId}/links/{linkFieldId}/records/{recordId}` API to list linked records for a specific record.

Let me know if this is enough for your use-case, otherwise I can help with manually recognizing the relations as well (for quick overview all information stored within columns part of table read API, and relation information is stored within colOptions of your column (mm table, child/parent columns etc.). we provide API's to fetch linked records without bothering with all the underlying stuff so I hope that would be enough for your use-case)

**Reply by sclu1034** [reply]
> If you would like to see some specific information related to linked records you can use a Lookup column ([docs.nocodb.com/fields/field-types/links-based/lookup](https://docs.nocodb.com/fields/field-types/links-based/lookup)) or Rollup column for aggregated information ([docs.nocodb.com/fields/field-types/links-based/rollup](https://docs.nocodb.com/fields/field-types/links-based/rollup/)).

Does adding a new, separate column to the table augment the API response in a way that I can figure out the internal name of an existing column? That'd be a rather unorthodox design.
To clarify: I don't need anything changed within the UI, the table is already set up the way I want it. I'm completely focused on the HTTP API right now.

> If you would like to get linked records via API you can use `/api/v2/tables/{tableId}/links/{linkFieldId}/records/{recordId}` API to list linked records for a specific record.

The response from `/api/v2/tables/{tableId}/records` already contains the data I want, I just need a way to (programmatically) figure out the column name that corresponds to the M2M relation so I can access it in the returned JSON.
If it's easier to figure out a `linkFieldId` than an internal column name, I could go that route. But as of now, I don't have that, either.

**Reply by sclu1034** [reply]
From `curl /api/v2/tables/{tableId}/records` I get:

```json
{
  "list": [
    {
      "Id": 1,
      "Name": "Sync Test",
      "CreatedAt": "2024-03-12 09:31:45+00:00",
      "UpdatedAt": "2024-03-12 11:11:34+00:00",
      "Songs": 2,
      "nc_o9qt___nc_m2m_odp0l09uz_s": [
        {
          "table2_id": 1,
          "table1_id": 1,
          "Playlists": {
            "Songs": 2,
            "Id": 1,
            "Name": "Sync Test",
            "CreatedAt": "2024-03-12 09:31:45+00:00",
            "UpdatedAt": "2024-03-12 11:11:34+00:00",
          }
        },
        {
          "table2_id": 1,
          "table1_id": 3,
          "Playlists": {
            "Songs": 2,
            "Id": 1,
            "Name": "Sync Test",
            "CreatedAt": "2024-03-12 09:31:45+00:00",
            "UpdatedAt": "2024-03-12 11:11:34+00:00",
          }
        }
      ]
    }
  ]
// pageInfo omitted
}
```

How do I determine that the field corresponding to `"Songs"` is called `"nc_o9qt___nc_m2m_odp0l09uz_s"`?

**Reply by pranavxc** [reply]
Apologies for any confusion caused. The `nc_o9qt___nc_m2m_odp0l09uz_s` field does not relate to the many-to-many (MM) relationship between Song and PlayList. It represents a "has-many" relationship between Playlist and a junction table, the junction table is used to achieve the Many to Many relation between Playlist and Songs table.  To bypass these system columns, you can include viewId as a query parameter in your request.

Regarding the Links column functionality, it operates similarly to a Rollup with a count function, providing the number of records associated with the linked table(`"Songs": 2,`). If you aim to retrieve the list of linked records for a specific row, you should utilize the following API endpoint:

https://data-apis-v2.nocodb.com/#tag/Table-Records/operation/db-data-table-row-nested-list

```
/api/v2/tables/{tableId}/links/{linkFieldId}/records/{recordId}
```
To get the `linkField` you can use our table read meta API - https://meta-apis-v2.nocodb.com/#tag/DB-Table/operation/db-table-read

FYI: And we are updating our APIs to always exclude these system columns from the API response.

**Reply by sclu1034** [reply]
Thanks for the additional details @pranavxc.

But looking at `/api/v2/tables/{tableId}/links/{linkFieldId}/records/{recordId}`, I don't get full records, but only the `"Id"` and `"Name"` fields. So despite doing an extra call per each record, I basically get the same info I already have in `"nc_o9qt___nc_m2m_odp0l09uz_s"`, because I'd still need extra calls to get the full data about the linked records in either case.

If that "internal" column is supposed to disappear, you might want to consider an explicit column listing all the linked IDs as part of `/api/v2/tables/{tableId}/records`. Because then you can get the ability to cross-reference IDs after just two calls to `/api/v2/tables/{tableId}/records`, rather than having to run lots of individual `/api/v2/tables/{tableId}/links/{linkFieldId}/records/{recordId}` calls.


Also, what purpose does the `"Songs": 2` serve? I can imagine a boolean of "has links" vs "has no links" to bypass additional logic to query those links, but what's a use case where you need to know "there are 17 links specifically", without also knowing details about them?

---

As a side note: Up until now I didn't know these two sites you just linked, despite having dug around `docs.nocodb.com` quite a lot. And I also stumbled upon `apis.nocodb.com` as a third one at some point.

I think you should consider adding a prominent (i.e. should be among the top results when searching for "API") page to `docs.nocodb.com` that
- lists all these sites that one should know/check
- explains how "meta" vs "data" sites interact (or maybe consolidate them, given that the "meta" site already has a "DATA APIS" section)
- for good measure, also lists the out-of-date sites to make sure people know what not to use
  - I'm still not sure if `apis.nocodb.com` is considered up-to-date given the existence of "v2" sites
  - for a while I was looking at a list of API routes on `docs.nocodb.com` until finally I noticed there was a version switch and that it was set to "Old UI". So I switched to "Latest", but then the page was gone with no alternative in sight -> you'll want to make sure users know where to check instead

**Reply by pranavxc** [reply]
> But looking at /api/v2/tables/{tableId}/links/{linkFieldId}/records/{recordId}, I don't get full records, but only the "Id" and "Name" fields.

By default in list we only includes primary key and display value but you can get all fields/limited fields by defining `fields` query parameter.

For example: `fields=*` for getting all columns or `fields=col1,col2,col3` for getting `col1` , `col2` and `col3`

> If that "internal" column is supposed to disappear, you might want to consider an explicit column listing all the linked IDs as part of /api/v2/tables/{tableId}/records.

We used to have this feature earlier and the type called `LinkToAnotherRecord`, we dropped that type and introduced `Links` type. The main reason is when there are large amount of links getting all the links with the root data will be too much, so we used to limit `25` but pagination of such data(nested link array) is difficult as well. Other than that the performance of list API improved after changing it to `Links` since we are just getting the count.

If you want to access any column from the linked table along with list API then you can create a Lookup column which extract that particular column information and returns as an array.

> As a side note: Up until now I didn't know these two sites you just linked, despite having dug around docs.nocodb.com quite a lot. And I also stumbled upon apis.nocodb.com as a third one at some point.

> I think you should consider adding a prominent (i.e. should be among the top results when searching for "API") page to docs.nocodb.com that
> 
> lists all these sites that one should know/check
> explains how "meta" vs "data" sites interact (or maybe consolidate them, given that the "meta" site already has a "DATA APIS" section)
> for good measure, also lists the out-of-date sites to make sure people know what not to use
> I'm still not sure if apis.nocodb.com is considered up-to-date given the existence of "v2" sites
> for a while I was looking at a list of API routes on docs.nocodb.com until finally I noticed there was a version switch and that it was set to "Old UI". So I switched to "Latest", but then the page was gone with no alternative in sight -> you'll want to make sure users know where to check instead

Thanks for pointing out, we will update these and make these links available in the docs and wherever necessary.

> Also, what purpose does the "Songs": 2 serve? I can imagine a boolean of "has links" vs "has no links" to bypass additional logic to query those links, but what's a use case where you need to know "there are 17 links specifically", without also knowing details about them?

It helps to get the identify how many total records are linked and based on that we can get child list with pagination.

**Reply by sclu1034** [reply]
> For example: `fields=*` for getting all columns

Aha, another goodie I haven't seen mentioned in the docs. It doesn't help with the "per record" overhead, though (see example below).

> The main reason is when there are large amount of links getting all the links with the root data will be too much, so we used to limit `25` but pagination of such data(nested link array) is difficult as well. Other than that the performance of list API improved after changing it to `Links` since we are just getting the count.

I see. But would that performance hit still apply if the data was just a list of IDs (e.g. `"Songs": [1, 5, 123, 234, ...]`)?

Right now, I could do everything with a fixed number of two parallel (paginated) API calls (`O(1)`), regardless of set size:

```js
// This is what I need to figure out deterministically.
// But with the change suggested above, this could just be "Songs"
const column_name = "nc_o9qt___nc_m2m_odp0l09uz_s";

// Two parallel HTTP calls
let [songs, playlists] = await Promise.all([
	get_list_records("Songs"),
	get_list_records("Playlists"),
]);

playlists = playlists.map(list => {
    // No further HTTP calls required to map the ID to the list we got above
    list.Songs = list[column_name].map(t => find_by_id(songs, t.table1_id));
    return list;
});
```

Whereas the variant via `linkFieldId` requires API calls based on the number of records in `Playlists` (`O(n)`), and is partially sequential:

```js
const field_id = /* get from table meta */;

// First HTTP call
let playlists = await get_list_records("Playlists");
// We might even have to chunk this, to avoid sending dozens of queries in parallel
playlists = await Promise.map(playlists, async list => {
	// One HTTP call per each `list`
	list.Songs = await get_linked_records(field_id, list.id, { fields = '*' });
	return list;
});
```
